### PR TITLE
Topic recognition - VQVAE on CSIRO HipMRI Dataset (47428421)

### PR DESCRIPTION
**Pull request by EwanStanich**
This pull request for for my implementation of a Vector Quantised Variational AutoEncoder (VQVAE) trained on the CSIRO HipMRI dataset. My repository contains all relevant files for the models, loading data, training the VQVAE and producing output. There are also images relevant to the README markdown file. The model achieves its criteria of a SSIM score greater than 0.6, with the highest achieved score being roughly 0.8. 

The files included are:
modules.py - contains the models for this project, including the Residual Layer, Residual Stack, Vector Quantiser, and the VQVAE. 

dataset.py - contains the functions used to load in the data from the specified folders containing Nifti files. 

train.py - contains the main training loop for the project, initializing all required models, folders, and optimizers before running through the training process. Hyperparameters are also specified at the top of this file for ease of changing. 

test.py - contains a code to generate images from the VQVAE. 

utils.py.- contains helper functions that assist with loading data, managing output folders, initialising models, calculating SSIM scores and plotting metrics. 

README.md – contains a in depth description of the model, providing example input and output as well as justifications for decisions made in the process of creating the model and steps to reproduce results. 

markdown_images – contains the images required for the README file.

For any further information on the codebase, refer to the README file which has more in-depth explanations of the project. 

Thank you,
Ewan Stanich 47428421

I’m very sorry about the number of commits made to try and fix my README file, I was having merge conflict issues and was making lots of changes trying to fix it as I can only tell if the fixes worked when they were live on Github.

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Very clear and resembling reconstructed images, with an SSIM over 0.8 (5 Marks)

2. Algorithm demonstrated and functions: (3 Marks)

4. Design:
Modularized design and good practice using utils script (1 Mark)

5. Commenting:
Sufficient docstrings and detailed comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:

- In README.md you could briefly justify the hyperparameters chosen for the training process, or any tuning on them (e.g. why ended up choosing LR of 3e-4, what were the other attempts and their implications), as well as the loss/SSIM plots;
- Images not shown properly on README.md due to incorrect path configuration;
- It seems you sampled the validation set from the test set, but there is another available validation set on Rangpur;
- It seems the loss was calculated for training and SSIM was calculated for validation, while both can be calculated for both training and validation for comparison.

**Comment by Uto0op** [comment]
Images loaded properly on README.md now.

**Comment by YashMitttal** [comment]
Some additional feedback:

- It’s great that you’ve specified the hyperparameters at the start of train.py, but it would be useful to provide a brief explanation of why these specific hyperparameters were chosen.
- Add a bit more detail on how you validated that the data was correctly processed (e.g., checks for image size, normalisation steps)
- Add references wherever required please.

Good job overall.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. No references.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-1</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (README). Please update for merge, doesn't affect grade.
